certora/spec/CVL2/timelockAuthorizerSasha.spec

```diff
@@ -0,0 +1,298 @@
+}
+
+
+// STATUS - verified
+// Only one flag executed/cancelled can be changed at a time
+rule onlyOneExecuteOrCancelCanChangeAtTime(env e, method f) {
+    uint256 actionIndex1; uint256 actionIndex2;
+    bool isExecuted1Before = getSchedExeExecuted(actionIndex1);
+    bool isExecuted2Before = getSchedExeExecuted(actionIndex2);
+    bool isCancelled1Before = getSchedExeCancelled(actionIndex1);
+    bool isCancelled2Before = getSchedExeCancelled(actionIndex2);
+
+    require actionIndex1 != actionIndex2;
+    require getSchedExeLength() < max_uint / 4;
+
+    calldataarg args;
+    f(e, args);
+
+    bool isExecuted1After = getSchedExeExecuted(actionIndex1);
+    bool isExecuted2After = getSchedExeExecuted(actionIndex2);
+    bool isCancelled1After = getSchedExeCancelled(actionIndex1);
+    bool isCancelled2After = getSchedExeCancelled(actionIndex2);
+
+    assert isExecuted1Before != isExecuted1After 
+                => (isExecuted2Before == isExecuted2After 
+                    && isCancelled1Before == isCancelled1After 
+                    && isCancelled2Before == isCancelled2After)
+                    && !isExecuted1Before;
+    assert isCancelled1Before != isCancelled1After 
+                => (isExecuted1Before == isExecuted1After 
+                    && isExecuted2Before == isExecuted2After 
+                    && isCancelled2Before == isCancelled2After)
+                    && !isCancelled1Before;
+    assert isExecuted2Before != isExecuted2After 
+                => (isExecuted1Before == isExecuted1After 
+                    && isCancelled1Before == isCancelled1After 
+                    && isCancelled2Before == isCancelled2After)
+                    && !isExecuted2Before;
+    assert isCancelled2Before != isCancelled2After 
+                => (isExecuted1Before == isExecuted1After 
+                    && isExecuted2Before == isExecuted2After 
+                    && isCancelled1Before == isCancelled1After)
+                    && !isCancelled2Before;
+}
+
+
+// STATUS - verified
+// Flag was changed by correct function. ("only one flag can be changed at a time" was checked above)
+rule onlyExecuteAndCancelCanChangeTheirFlags(env e, method f) {
+    uint256 actionIndex1; uint256 actionIndex2;
+    bool isExecuted1Before = getSchedExeExecuted(actionIndex1);
+    bool isCancelled1Before = getSchedExeCancelled(actionIndex1);
+
+    require actionIndex1 != actionIndex2;
+    require limitArrayLength();
+
+    calldataarg args;
+    f(e, args);
+
+    bool isExecuted1After = getSchedExeExecuted(actionIndex1);
+    bool isCancelled1After = getSchedExeCancelled(actionIndex1);
+
+    assert (isExecuted1Before != isExecuted1After)
+            => f.selector == sig:execute(uint256).selector;
+    assert (isCancelled1Before != isCancelled1After)
+            => f.selector == sig:cancel(uint256).selector;
+}
+
+
+// STATUS - verified
+// if canExecute returns true / false, execute finishes successfully / reverts.
```

Comment:
because it's not completely true imo
`execute()` has an additional `if` that, if true triggers another require. Also, the tool will find a case when it can revert because `msg.value > 0`

Reply:
The rule description should then be updated to something like:
`An execute cannot succeed if canExecute is false`

Reply:
Oh, ok I got your point. Actually, it was described accordingly to implementation in the report. Seems like I forgot to update the spec.